h/static/scripts/base/controller.js

```diff
@@ -86,6 +106,95 @@ class Controller {
   forceUpdate() {
     this.update(this.state, this.state);
   }
+
+  /**
+   * Replace the HTML content of the element with an updated version from the
+   * server.
+   *
+   * Since this replaces the entire element, the current controller will not
+   * receive events for the replaced element. Instead the new controller for the
+   * replaced element is returned and can be used to transfer state from the
+   * existing controller.
+   *
+   * @param {string} html - The new markup for the element
+   * @return {Controller} - The new controller instance of the same type as this
+   *         controller on the element that replaces `this.element`
+   */
+  reload(html) {
+    if (!this.options.reload) {
+      throw new Error('No reload() function supplied to controller constructor');
+    }
+
+    var newElement = this.options.reload(this.element, html);
+    var ctrl = (newElement.controllers || []).find(ctrl => {
+      return ctrl instanceof this.constructor;
+    });
+
+    this.trigger('reload', {newController: ctrl});
+
+    return ctrl;
+  }
+
+  /**
+   * Return controllers attached to child elements which are instances of
+   * `ControllerClass`
+   *
+   * @return {Array<ControllerClass>} Array of controllers attached to children
+   *         of `this.element` which are instances of `ControllerClass`
+   */
+  childControllers(ControllerClass) {
+    var children = this.element.querySelectorAll('*');
+    var controllers = [];
+    for (var i=0; i < children.length; i++) {
+      var child = children[i];
+      if (!child.controllers) {
+        continue;
+      }
+      child.controllers.forEach(ctrl => {
+        if (ctrl instanceof ControllerClass) {
+          controllers.push(ctrl);
+        }
+      });
+    }
+    return controllers;
+  }
+
+  /**
+   * Add a listener for a given DOM event emitted either on `this.element` or
+   * events that bubble up from children.
+   */
+  on(event, listener) {
+    this.element.addEventListener(event, listener);
+    this._listeners.push({event, listener});
+  }
+
+  /**
+   * Broadcast an event to event handlers registered on the current element and
+   * parent elements using `on`.
+   *
+   * @param {string|Event} event - The name of an event, or the event to emit
+   * @param {Object} [data] - Optional data to attach to the event as `event.data`
+   */
+  trigger(event, data) {
+    if (typeof event === 'string') {
+      event = new Event(event);
```

Comment:
It's not a good idea to change the type of a variable halfway through a function. Let's call the argument here `eventType`.